jormungandr/src/rest/mod.rs

```diff
@@ -58,19 +68,62 @@ pub async fn start_rest_server(config: Rest, explorer_enabled: bool, context: Co
                 }
             }
             span
-        }))
-        .or(prometheus::filter(context.clone()));
-    if explorer_enabled {
+        }));
+
+    setup_prometheus(api, config, context, stopper_rx).await;
+}
+
+#[cfg(feature = "prometheus-metrics")]
+async fn setup_prometheus<App>(
+    app: App,
+    config: Config,
+    context: ContextLock,
+    shutdown_signal: impl Future<Output = ()> + Send + 'static,
+) where
+    App: Filter<Error = warp::Rejection> + Clone + Send + Sync + 'static,
+    App::Extract: warp::Reply,
+{
+    if config.enable_prometheus {
+        let prometheus = prometheus::filter(context.clone());
+        setup_explorer(app.or(prometheus), config, context, shutdown_signal).await;
+    } else {
+        setup_explorer(app, config, context, shutdown_signal).await;
+    }
+}
+
+#[cfg(not(feature = "prometheus-metrics"))]
+async fn setup_prometheus<App>(
+    app: App,
+    config: Config,
+    context: ContextLock,
+    shutdown_signal: impl Future<Output = ()> + Send + 'static,
+) where
+    App: Filter<Error = warp::Rejection> + Clone + Send + Sync + 'static,
+    App::Extract: warp::Reply,
+{
+    setup_explorer(app, config, context, shutdown_signal).await;
+}
+
+async fn setup_explorer<App>(
+    app: App,
+    config: Config,
+    context: ContextLock,
+    shutdown_signal: impl Future<Output = ()> + Send + 'static,
+) where
+    App: Filter<Error = warp::Rejection> + Clone + Send + Sync + 'static,
+    App::Extract: warp::Reply,
+{
+    if config.enable_explorer {
         let explorer = explorer::filter(context);
-        setup_cors(api.or(explorer), config, stopper_rx).await;
+        setup_cors(app.or(explorer), config, shutdown_signal).await;
     } else {
-        setup_cors(api, config, stopper_rx).await;
+        setup_cors(app, config, shutdown_signal).await;
```

Comment:
I don't like how `setup_cors` is now called inside of one of the app setup functions.
Can we make each filter setup function return an `impl Filter + ...` to be able to chain them like warp intended?

Reply:
Tried it, and it's not actually easy without lots of boxing.